File: Lab/Cryptography and Network Security Lab/Statistical and Multiplicative.py
import sys
from collections import Counter

# 1. GCD Function
def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

# 2. Modular Inverse Function (Decryption Key বের করার জন্য)
def mod_inverse(k):
    for i in range(1, 26):
        if (k * i) % 26 == 1:
            return i
    return -1

# 3. Encryption Function
def encrypt(plaintext, key):
    cipher = ""
    for c in plaintext:
        if c.isalpha():
            p = ord(c.lower()) - ord('a')
            c_val = (p * key) % 26
            cipher += chr(c_val + ord('a'))
        else:
            cipher += c
    return cipher

# 4. Decryption Function
def decrypt(cipher, key):
    inv = mod_inverse(key)
    if inv == -1: return "Invalid Key"
    
    plain = ""
    for c in cipher:
        if c.isalpha():
            c_val = ord(c.lower()) - ord('a')
            p = (c_val * inv) % 26
            plain += chr(p + ord('a'))
        else:
            plain += c
    return plain

# 5. Statistical Attack (With Human Verification & Decryption Key Display)
def statistical_attack_interactive(cipher, sample_text):
    print("\n--- Statistical Attack (Interactive Mode) ---")
    
    # সাইফারের ফ্রিকোয়েন্সি (Top 3)
    cipher_chars = [c.lower() for c in cipher if c.isalpha()]
    if not cipher_chars: return
    top_cipher = Counter(cipher_chars).most_common(3)
    
    # স্যাম্পলের ফ্রিকোয়েন্সি (Top 3)
    sample_chars = [c.lower() for c in sample_text if c.isalpha()]
    if not sample_chars: return
    top_sample = Counter(sample_chars).most_common(3)

    print(f"Top Cipher Chars: {[x[0] for x in top_cipher]}")
    print(f"Top Sample Chars: {[x[0] for x in top_sample]}")
    print("Checking hypotheses... (Type 'y' if the text makes sense, 'n' to continue)\n")

    # ৩x৩ লুপ চালানো
    for c_item in top_cipher:
        for s_item in top_sample:
            
            cipher_char = c_item[0]
            sample_char = s_item[0]
            
            c_val = ord(cipher_char) - ord('a')
            p_val = ord(sample_char) - ord('a')

            # Plaintext 'a' (value 0) হলে স্কিপ
            if p_val == 0:
                continue

            # K খোঁজা
            found_k = -1
            for k in range(1, 26):
                if gcd(k, 26) == 1:
                    if (p_val * k) % 26 == c_val:
                        found_k = k
                        break 
            
            if found_k != -1:
                # ডিক্রিপশন কী (Inverse Key) বের করা হচ্ছে
                decryption_key = mod_inverse(found_k) 
                decrypted_preview = decrypt(cipher, found_k)
                
                print(f"Hypothesis: Cipher '{cipher_char}' -> Sample '{sample_char}'")
                print(f"  -> Encryption Key (K): {found_k}")
                print(f"  -> Decryption Key (Inverse): {decryption_key}") # এখানে শো করা হচ্ছে
                print(f"  -> Decrypted Text: {decrypted_preview}")
                
                # এখানে ইউজার ইনপুট নেওয়া হবে
                user_check = input("  >>> Is this the correct text? (y/n): ").lower()
                
                if user_check == 'y':
                    print("\n[SUCCESS] Attack Successful! Keys confirmed.")
                    return 
                else:
                    print("  -> Continuing search...\n")

    print("\n[FAILED] No suitable key accepted by user in top 3x3 combinations.")

def main():
    plaintext = input("Enter plaintext: ")
    try:
        key = int(input("Enter Encryption key (must be coprime to 26): "))
    except ValueError: return

    if gcd(key, 26) != 1:
        print("Invalid Key!")
        return

    print("\n--- For Statistical Attack ---")
    print("Enter a sample text:")
    sample_text = input()
    if len(sample_text) < 1: sample_text = "eeee tttt aaaa oooo iiii"

    cipher = encrypt(plaintext, key)
    print(f"\nEncrypted text: {cipher}")
    
    # ইন্টারেক্টিভ ফাংশন কল
    statistical_attack_interactive(cipher, sample_text)

if __name__ == "__main__":
    main()

File: Lab/Cryptography and Network Security Lab/Statistical and Multiplicative.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

// GCD বের করার ফাংশন
int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Multiplicative Inverse বের করার ফাংশন (X বের করা)
int modInverse(int k) {
    for (int i = 1; i < 26; i++) {
        if ((k * i) % 26 == 1)
            return i;
    }
    return -1;
}

// এনক্রিপশন ফাংশন
string encrypt(string plaintext, int key) {
    string cipher = "";
    for (char c : plaintext) {
        if (isalpha(c)) {
            int p = tolower(c) - 'a';
            int ciph = (p * key) % 26;
            cipher += char(ciph + 'a'); // আউটপুট আপারকেস হবে
        } else {
            cipher += c; // স্পেস বা অন্য ক্যারেক্টার অপরিবর্তিত থাকবে
        }
    }
    return cipher;
}

// ডিক্রিপশন ফাংশন (Ciphertext এবং Key ব্যবহার করে)
string decrypt(string cipher, int key) {
    int inv = modInverse(key);
    if (inv == -1) return "Invalid Key for Decryption";

    string plain = "";
    for (char c : cipher) {
        if (isalpha(c)) {
            int ciph = toupper(c) - 'A'; // ইনপুট আপারকেস ধরে নেওয়া হচ্ছে
            int p = (ciph * inv) % 26;
            plain += char(p + 'a'); // আউটপুট লোয়ারকেস হবে
        } else {
            plain += c;
        }
    }
    return plain;
}

// Statistical Attack ফাংশন (Top 3 Frequent Character চেক করবে)
void statisticalAttack(string cipher) {
    cout << "\n--- Statistical Attack Results ---" << endl;

    // ১. ফ্রিকোয়েন্সি ম্যাপ তৈরি
    map<char, int> freqMap;
    for (char c : cipher) {
        if (isalpha(c)) {
            freqMap[toupper(c)]++;
        }
    }

    if (freqMap.empty()) {
        cout << "No letters found to analyze." << endl;
        return;
    }

    // ২. ফ্রিকোয়েন্সি অনুযায়ী সর্ট করা (Descending Order)
    vector<pair<char, int>> sortedFreq(freqMap.begin(), freqMap.end());
    sort(sortedFreq.begin(), sortedFreq.end(),
         [](const pair<char, int>& a, const pair<char, int>& b) {
             return a.second > b.second; // বেশি ফ্রিকোয়েন্সি আগে থাকবে
         });

    // ৩. Top 3 ক্যারেক্টার দিয়ে লুপ চালানো
    int limits = min((int)sortedFreq.size(), 3);

    for (int i = 0; i < limits; i++) {
        char cipherChar = sortedFreq[i].first;
        int c_val = cipherChar - 'A';
        int p_val = 4; // ধরে নিচ্ছি প্লেইনটেক্সট 'e' (যার ভ্যালু 4)

        cout << "\nAttempt " << i + 1 << ": Assuming '" << cipherChar << "' maps to 'e' (4)" << endl;

        // ৪. K বের করা: (4 * K) % 26 == c_val
        int foundK = -1;
        for (int k = 1; k < 26; k++) {
            // Key অবশ্যই 26 এর সাথে কো-প্রাইম হতে হবে
            if (gcd(k, 26) == 1) {
                if ((p_val * k) % 26 == c_val) {
                    foundK = k;
                    break;
                }
            }
        }

        if (foundK != -1) {
            // ৫. K পাওয়া গেলে Inverse (X) বের করা এবং ডিক্রিপ্ট করা
            int inverseX = modInverse(foundK);
            cout << "  -> Found Potential Key (K): " << foundK << endl;
            cout << "  -> Multiplicative Inverse (X): " << inverseX << endl;
            cout << "  -> Decrypted Message: " << decrypt(cipher, foundK) << endl;
        } else {
            cout << "  -> No valid key found for this assumption (Equation 4*K = " << c_val << " mod 26 has no solution)." << endl;
        }
    }
}

int main() {
    string plaintext;
    int key;

    // ইনপুট নেওয়া
    cout << "Enter plaintext: ";
    getline(cin, plaintext);

    cout << "Enter key (must be coprime to 26): ";
    cin >> key;

    // ভ্যালিডেশন চেক
    if (gcd(key, 26) != 1) {
        cout << "Invalid Key! GCD(key, 26) must be 1." << endl;
        return 0;
    }

    // এনক্রিপশন
    string cipher = encrypt(plaintext, key);
    cout << "\nEncrypted text: " << cipher << endl;

    // সাধারণ ডিক্রিপশন (চেক করার জন্য)
    cout << "Decrypted with original key: " << decrypt(cipher, key) << endl;

    // স্ট্যাটিসটিক্যাল অ্যাটাক চালানো
    statisticalAttack(cipher);

    return 0;
}
